basestar-storage/src/main/java/io/basestar/storage/util/KeysetPagingUtils.java

```diff
@@ -80,19 +75,81 @@ private KeysetPagingUtils() {
 
-        try(final ByteArrayOutputStream baos = new ByteArrayOutputStream();
-            final DataOutputStream dos = new DataOutputStream(baos)) {
-            for (final Sort s : sort) {
-                final Name name = s.getName();
-                final Use<Object> type = schema.typeOf(name);
-                final Object value = type.create(name.get(object));
-                type.serialize(value, dos);
-            }
+        try (final ByteArrayOutputStream baos = new ByteArrayOutputStream();
+             final DataOutputStream dos = new DataOutputStream(baos)) {
+            addSortingInfo(schema, sort, object, dos);
+            dos.flush();
+            return new Page.Token(baos.toByteArray());
+        } catch (final IOException e) {
+            // Shouldn't be possible, not doing real IO
+            throw new UncheckedIOException(e);
+        }
+    }
+
+    public static Page.Token countPreservingKeysetPagingToken(final InstanceSchema schema,
+                                                              final List<Sort> sort,
+                                                              final Map<String, Object> object,
+                                                              final long count) {
+
+        try (final ByteArrayOutputStream baos = new ByteArrayOutputStream();
+             final DataOutputStream dos = new DataOutputStream(baos)) {
+            final UseInteger countType = UseInteger.DEFAULT;
+            countType.serialize(count, dos);
+            addSortingInfo(schema, sort, object, dos);
+
             dos.flush();
             return new Page.Token(baos.toByteArray());
         } catch (final IOException e) {
             // Shouldn't be possible, not doing real IO
             throw new UncheckedIOException(e);
         }
     }
+
+
+    public static Map<String, List<Object>> countPreservingKeysetValues(final InstanceSchema schema,
+                                                                        final List<Sort> sort,
+                                                                        final Page.Token token) {
+        final byte[] bytes = token.getValue();
+        final List<Object> values = new ArrayList<>();
+        final HashMap<String, List<Object>> results = new HashMap<>();
+        long count;
+        try (final ByteArrayInputStream bais = new ByteArrayInputStream(bytes);
+             final DataInputStream dis = new DataInputStream(bais)) {
+
+            final UseInteger countType = UseInteger.DEFAULT;
+            count = countType.deserialize(dis);
+            getSortingInfo(schema, sort, dis, values::add);
+        } catch (final IOException e) {
+            // Shouldn't be possible, not doing real IO
+            throw new UncheckedIOException(e);
+        }
+        assert values.size() == sort.size();
+        results.put("sort", values);
+        results.put("count", Collections.singletonList(count));
```

Comment:
what is the behaviour if the `token` does not contain `count` information? E.g. mangled/spoofed token, or if Total was not requested in the initial call? Does it throw a proper exception? Or a 500 error?

Reply:
Currently all malformed tokens will 500 - there is not a 4xx mapped exception class for token problems, we could add one though, if you add a new `class PagingTokenSyntaxException extends RuntimeException implements io.basestar.exception.HasExceptionMetadata`, you can set the response code for that exception to e.g. 400